File: src/sugtao4423/lod/dialog_onclick/Dialog_ListClick.java
package sugtao4423.lod.dialog_onclick;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import twitter4j.MediaEntity;
import twitter4j.Status;
import twitter4j.User;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.DialogInterface.OnClickListener;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.GridLayout;
import android.widget.LinearLayout;
import android.widget.AdapterView.OnItemClickListener;
import sugtao4423.lod.ChromeIntent;
import sugtao4423.lod.IntentActivity;
import sugtao4423.lod.Keys;
import sugtao4423.lod.ListViewListener;
import sugtao4423.lod.R;
import sugtao4423.lod.ShowToast;
import sugtao4423.lod.Show_Video;
import sugtao4423.lod.swipe_image_viewer.ImageFragmentActivity;
import sugtao4423.lod.tweetlistview.TweetListAdapter;
import sugtao4423.lod.tweetlistview.TweetListView;
import sugtao4423.lod.userpage_fragment.UserPage;
import sugtao4423.lod.utils.Regex;
import sugtao4423.lod.utils.Utils;

public class Dialog_ListClick implements OnItemClickListener{

	private Context context;
	private Status status;
	private ArrayList<Status> listData;
	private AlertDialog dialog;

	public Dialog_ListClick(Context context, Status status, ArrayList<Status> listData, AlertDialog dialog){
		this.context = context;
		this.status = status;
		this.listData = listData;
		this.dialog = dialog;
	}

	@Override
	public void onItemClick(AdapterView<?> parent, View view, int position, long id){
		dialog.dismiss();
		String clickedText = (String)parent.getItemAtPosition(position);

		if(clickedText.equals("正規表現で抽出")){
			View regView = View.inflate(context, R.layout.reg_dialog, null);
			final EditText regEdit = (EditText)regView.findViewById(R.id.regDialog_edit);
			GridLayout gridLayout = (GridLayout)regView.findViewById(R.id.regDialog_grid);
			String[] regItems = new String[]{".", "*", "|", "+", "?", "\\", "^", "$", "(", ")", "[", "]", "{", "}"};
			LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(130, 130);
			for(int i = 0; i < regItems.length; i++){
				Button btn = new Button(context);
				btn.setLayoutParams(params);
				btn.setText(regItems[i]);
				btn.setOnClickListener(new Dialog_regButtonClick(regEdit));
				gridLayout.addView(btn);
			}

			final SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
			regEdit.setText(pref.getString(Keys.REGULAR_EXPRESSION, ""));
			new AlertDialog.Builder(context)
			.setTitle("正規表現を入力してください")
			.setView(regView)
			.setNegativeButton("キャンセル", null)
			.setPositiveButton("OK", new OnClickListener(){
				@Override
				public void onClick(DialogInterface dialog, int which){
					String editReg = regEdit.getText().toString();
					pref.edit().putString(Keys.REGULAR_EXPRESSION, editReg).commit();
					Pattern pattern = Pattern.compile(editReg, Pattern.DOTALL);
					TweetListAdapter adapter = new TweetListAdapter(context);
					boolean exists = false;
					for(Status s : listData){
						if(pattern.matcher(s.getText()).find()){
							adapter.add(s);
							exists = true;
						}
					}
					if(!exists){
						new ShowToast(R.string.nothing, context, 0);
					}else{
						TweetListView l = new TweetListView(context);
						l.setAdapter(adapter);
						adapter.setOnItemClickListener(new ListViewListener());
						adapter.setOnItemLongClickListener(new ListViewListener());
						new AlertDialog.Builder(context).setView(l).show();
					}
				}
			}).show();
		}else if(clickedText.startsWith("http") || clickedText.startsWith("ftp")){
			Matcher image = Regex.media_image.matcher(clickedText);
			Matcher video = Regex.media_video.matcher(clickedText);
			Matcher gif = Regex.media_gif.matcher(clickedText);
			Matcher state = Regex.statusUrl.matcher(clickedText);
			Intent intent;
			if(image.find()){
				ArrayList<String> urls = new ArrayList<String>();
				MediaEntity[] mentitys = status.getMediaEntities();
				if(mentitys != null && mentitys.length > 0){
					for(MediaEntity media : mentitys){
						if(!Utils.isVideoOrGif(media))
							urls.add(media.getMediaURL());
					}
				}
				int pos = urls.indexOf(clickedText);
				String[] arr = (String[])urls.toArray(new String[0]);
				intent = new Intent(context, ImageFragmentActivity.class);
				intent.putExtra(ImageFragmentActivity.INTENT_EXTRA_KEY_URLS, arr);
				intent.putExtra(ImageFragmentActivity.INTENT_EXTRA_KEY_POSITION, pos);
			}else if(video.find()){
				intent = new Intent(context, Show_Video.class);
				intent.putExtra(Show_Video.INTENT_EXTRA_KEY_URL, clickedText);
				intent.putExtra(Show_Video.INTENT_EXTRA_KEY_TYPE, Show_Video.TYPE_VIDEO);
			}else if(gif.find()){
				intent = new Intent(context, Show_Video.class);
				intent.putExtra(Show_Video.INTENT_EXTRA_KEY_URL, clickedText);
				intent.putExtra(Show_Video.INTENT_EXTRA_KEY_TYPE, Show_Video.TYPE_GIF);
			}else if(state.find()){
				new IntentActivity().showStatus(Long.parseLong(state.group(2)), context, false);
				return;
			}else{
				new ChromeIntent(context, Uri.parse(clickedText));
				return;
			}
			context.startActivity(intent);
		}else if(clickedText.equals("ブラウザで開く")){
			Status orig = status.isRetweet() ? status.getRetweetedStatus() : status;
			String tweet_sn = orig.getUser().getScreenName();
			String tweet_id = String.valueOf(orig.getId());
			String url = "https://twitter.com/" + tweet_sn + "/status/" + tweet_id;
			context.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
		}else if(clickedText.startsWith("@")){ // UserPage
			Intent intent = new Intent(context, UserPage.class);
			String usersn = clickedText.substring(1);
			User user = status.getUser().getScreenName().equals(usersn) ? status.getUser() : null;
			if(user != null)
				intent.putExtra(UserPage.INTENT_EXTRA_KEY_USER_OBJECT, user);
			else
				intent.putExtra(UserPage.INTENT_EXTRA_KEY_USER_SCREEN_NAME, usersn);
			context.startActivity(intent);
		}
	}

	public void regEditPlus(EditText edit, String text){
		edit.setText(edit.getText().toString() + text);
	}
}

File: app/src/main/java/sugtao4423/lod/dialog_onclick/Dialog_ListClick.kt
package sugtao4423.lod.dialog_onclick

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import sugtao4423.lod.*
import sugtao4423.lod.swipe_image_viewer.ImageFragmentActivity
import sugtao4423.lod.tweetlistview.TweetListAdapter
import sugtao4423.lod.tweetlistview.TweetListView
import sugtao4423.lod.ui.userpage.UserPageActivity
import sugtao4423.lod.utils.Regex
import sugtao4423.lod.utils.Utils
import twitter4j.Status
import java.util.regex.Pattern
import java.util.regex.PatternSyntaxException

class Dialog_ListClick(private val status: Status, private val listData: ArrayList<Status>, private val dialog: AlertDialog) :
        AdapterView.OnItemClickListener {

    private lateinit var context: Context

    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        if (parent == null) {
            return
        }
        dialog.dismiss()
        context = parent.context
        val clickedText = parent.getItemAtPosition(position) as String

        when {
            clickedText == context.getString(R.string.extract_with_regex) -> extractTweetsWithRegex()
            clickedText.contains(Regex("^(http|ftp)")) -> openUrl(clickedText)
            clickedText == context.getString(R.string.open_in_browser) -> openInBrowser()
            clickedText.startsWith("@") -> openUserPage(clickedText.substring(1))
        }
    }

    private fun extractTweetsWithRegex() {
        val regView = View.inflate(context, R.layout.reg_dialog, null)
        val regEdit = regView.findViewById<EditText>(R.id.regDialogEdit)
        val regIncludeRT = regView.findViewById<CheckBox>(R.id.regDialogIncludeRetweet)
        val gridLayout = regView.findViewById<GridLayout>(R.id.regDialogGrid)
        val regItems = arrayOf(".", "*", "|", "+", "?", "\\", "^", "$", "(", ")", "[", "]", "{", "}")
        val params = LinearLayout.LayoutParams(130, 130)
        regItems.map {
            val btn = Button(context).apply {
                layoutParams = params
                text = it
                setOnClickListener(Dialog_regButtonClick(regEdit))
            }
            gridLayout.addView(btn)
        }

        val prefRepo = (context.applicationContext as App).prefRepository
        regEdit.setText(prefRepo.regularExpression)
        AlertDialog.Builder(context).also {
            it.setTitle(R.string.input_regex)
            it.setView(regView)
            it.setNegativeButton(R.string.cancel, null)
            it.setPositiveButton(R.string.ok) { _, _ ->
                val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(regEdit.windowToken, 0)

                val editReg = regEdit.text.toString()
                prefRepo.regularExpression = editReg
                val pattern: Pattern
                try {
                    pattern = Pattern.compile(editReg, Pattern.DOTALL)
                } catch (e: PatternSyntaxException) {
                    ShowToast(context.applicationContext, R.string.invalid_pattern)
                    return@setPositiveButton
                }
                val adapter = TweetListAdapter(context)
                adapter.addAll(listData.filter { status ->
                    pattern.matcher(status.text).find() &&
                            (regIncludeRT.isChecked || (!regIncludeRT.isChecked && !status.isRetweet))
                })
                if (adapter.itemCount == 0) {
                    ShowToast(context.applicationContext, R.string.nothing)
                } else {
                    val l = TweetListView(context)
                    l.adapter = adapter
                    AlertDialog.Builder(context).setView(l).show()
                    ShowToast(context.applicationContext, R.string.param_regex_result_count, listData.size, adapter.itemCount)
                }
            }
            it.show()
        }
    }

    private fun openUrl(urlText: String) {
        val image = Regex.mediaImage.matcher(urlText)
        val video = Regex.mediaVideo.matcher(urlText)
        val gif = Regex.mediaGif.matcher(urlText)
        val state = Regex.statusUrl.matcher(urlText)
        val intent = when {
            image.find() -> {
                val urls = arrayListOf<String>()
                status.mediaEntities.map {
                    if (!Utils.isVideoOrGif(it)) {
                        urls.add(it.mediaURLHttps)
                    }
                }
                val pos = urls.indexOf(urlText)
                Intent(context, ImageFragmentActivity::class.java).apply {
                    putExtra(ImageFragmentActivity.INTENT_EXTRA_KEY_URLS, urls.toTypedArray())
                    putExtra(ImageFragmentActivity.INTENT_EXTRA_KEY_POSITION, pos)
                }
            }
            video.find() -> {
                Intent(context, ShowVideo::class.java).apply {
                    putExtra(ShowVideo.INTENT_EXTRA_KEY_URL, urlText)
                    putExtra(ShowVideo.INTENT_EXTRA_KEY_TYPE, ShowVideo.TYPE_VIDEO)
                }
            }
            gif.find() -> {
                Intent(context, ShowVideo::class.java).apply {
                    putExtra(ShowVideo.INTENT_EXTRA_KEY_URL, urlText)
                    putExtra(ShowVideo.INTENT_EXTRA_KEY_TYPE, ShowVideo.TYPE_GIF)
                }
            }
            state.find() -> {
                Intent(context, IntentActivity::class.java).apply {
                    putExtra(IntentActivity.TWEET_ID, state.group(Regex.statusUrlStatusIdGroup)!!.toLong())
                }
            }
            else -> {
                ChromeIntent(context, Uri.parse(urlText))
                return
            }
        }
        context.startActivity(intent)
    }

    private fun openInBrowser() {
        val orig = if (status.isRetweet) status.retweetedStatus else status
        val tweetSn = orig.user.screenName
        val tweetId = orig.id.toString()
        val url = "https://twitter.com/$tweetSn/status/$tweetId"
        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    private fun openUserPage(userScreenName: String) {
        val intent = Intent(context, UserPageActivity::class.java)
        if (status.user.screenName == userScreenName) {
            intent.putExtra(UserPageActivity.INTENT_EXTRA_KEY_USER_OBJECT, status.user)
        } else {
            intent.putExtra(UserPageActivity.INTENT_EXTRA_KEY_USER_SCREEN_NAME, userScreenName)
        }
        context.startActivity(intent)
    }

}
